pkg/controller/association/reconciler.go

```diff
@@ -378,10 +378,16 @@ func (r *Reconciler) updateStatus(ctx context.Context, associated commonv1.Assoc
 	defer span.End()
 
 	oldStatus := associated.AssociationStatusMap(r.AssociationType)
+
+	// To correctly compare statuses without making the reconciler aware of singleton vs multiple associations status
+	// differences we: set new status, get it from associated and only then compare with the oldStatus. Setting the
+	// same status is harmless, setting a different status is fine as we have a copy of oldStatus above.
+	if err := associated.SetAssociationStatusMap(r.AssociationType, newStatus); err != nil {
```

Comment:
This has a smell to me. I wonder if it would be marginally better to have

```go
	SetAssociationStatusMap(typ AssociationType, statusMap AssociationStatusMap) (bool, error)
```
where `bool` would signal whether the status had changed compared to the old value, but I am OK with doing this in a later refactoring given that this PR is intended as a bug fix with minimal changeset.

Reply:
It might make it a bit clearer, but the call to get still isn't avoided - to pass in the correct `newStatus` for `annotation.ForAssociationStatusChange` and `r.recorder.AnnotatedEventf` it needs to come from a call to `associated`. Otherwise it comes from the `Reconcile` func which doesn't have the knowledge of single vs multi and always supplies (sometimes incorrectly) namespaced name.